accepted/2020/platform-checks/platform-checks.md

```diff
@@ -0,0 +1,450 @@
+                                         int major,
+                                         int minor,
+                                         int build,
+                                         int revision);
+    }
+
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=false, Inherited=true)]
+    public sealed class RemovedInOSPlatformAttribute : OSPlatformAttribute
+    {
+        public PlatformSpecificAttribute(string osPlatform,
+                                         int major,
+                                         int minor,
+                                         int build,
+                                         int revision);
+    }
+
+    [AttributeUsage(AttributeTargets.Assembly |
+                    AttributeTargets.Class |
+                    AttributeTargets.Constructor |
+                    AttributeTargets.Event |
+                    AttributeTargets.Method |
+                    AttributeTargets.Module |
+                    AttributeTargets.Property |
+                    AttributeTargets.Struct,
+                    AllowMultiple=false, Inherited=true)]
+    public sealed class ObsoletedInOSPlatformAttribute : OSPlatformAttribute
+    {
+        public PlatformSpecificAttribute(string osPlatform,
+                                         int major,
+                                         int minor,
+                                         int build,
+                                         int revision);
+        public string Url { get; set; }
+    }
+}
+```
+
+### Platform context
+
+To determine the platform context of the call site the analyzer must consider
+the following two configurations:
+
+1. **Call-site**. Application of `AddedInOSPlatformAttribute` to the containing
+   member, type, module, or assembly.
+
+2. **Project** The project TFMs and `TargetPlatformMinVersion`. If the project
+   has an platform-specific TFM (such as `net5.0-ios13.0`) we consider it to be
+   on the specified platform. If `TargetPlatformMinVersion` is set, we assume
+   that version, otherwise we assume the version as specified by the TFM.
+
+If (2) is platform neutral (e.g. `net5.0`), the platform context is entirely
+determined by (1).
+
+If (2) is platform-specific it is combined with (1):
+
+* If the platforms between and (2) are disjoint (such as iOS and Windows), the
+  information from (1) is discarded.
+* Otherwise the platform context's minimum version is the maximum of the
+  versions determined by (1) and (2).
+
+### Diagnostics
+
+The analyzer can produce three diagnostics:
+
+1. '{API}' requires {OS} {OS version} or later.
```

Comment:
What about cases where a particular OS is the exception?  For example, there are some APIs that work on Windows and Linux, but not on macOS.

Reply:
I used to think in those terms but I don't think that's useful because it gets complicated fast. I think we're better off modelling these as two distinct concepts:

1. Platform specific APIs. These are specific OS. Examples: Registry, WinForms, NS*.
2. Partially portable APIs. These are features that aren't OS specific but can only be implemented on some operating systems/execution environments. Examples: RefEmit, file system access, thread creation.

I think we can safely model (1) with OS checks because the set of supported OS is fixed and known a-priori. Burning metadata in the reference assemblies is fine. For (2) I think we're better of modelling them as capability APIs so that when the support matrix changes, less code is broken. We can try to formalize capability APIs as well and provide an analyzer for that (#111) but I consider that out of scope for this feature.

Reply:
I have also added a question to the Q&A section.

Reply:
I'm not following the logic here. How is it more complicated to say that API x is only supported on platform y than to say it's only supported on platform y and z, or everywhere but y? We do this all over the place in our tests, and it works fine. Before making up your mind here, please do a search across all of our tests for things like PlatformSpecific, paying particular attention to things that are "AnyUnix" (i.e. anything but Windows) or ~something.